drivers/usbc/tcpc/stm32_ucpd.c

```diff
@@ -0,0 +1,1426 @@
+
+	data->discharge_vbus_cb = discharge_vbus_cb;
+}
+
+/**
+ * @brief Checks if VBUS is at a particular level
+ *
+ * @return true if VBUS is at the level voltage, else false
+ */
+static bool ucpd_check_vbus_level(const struct device *dev,
+				  enum tc_vbus_level level)
+{
+	struct tcpc_data *data = dev->data;
+	int vbus_meas;
+
+	if (data->vbus_cb == NULL) {
+		return false;
+	}
+
+	/* Call user supplied callback to measure VBUS */
+	vbus_meas = data->vbus_cb(dev);
+
+	switch (level) {
+	case TC_VBUS_SAFE0V:
+		return (vbus_meas < PD_V_SAFE_0V_MAX_MV);
+	case TC_VBUS_PRESENT:
+		return (vbus_meas >= PD_V_SAFE_5V_MIN_MV);
+	case TC_VBUS_REMOVED:
+		return (vbus_meas < TC_V_SINK_DISCONNECT_MAX_MV);
+	}
+
+	return false;
+}
+
+/**
+ * @brief Reads and returns VBUS measured in mV
+ *
+ * @return 0 on success
+ * @return -EIO on failure
+ * @return -EFAULT on buf being NULL
+ */
+static int ucpd_get_vbus(const struct device *dev, int *vbus_meas)
+{
+	struct tcpc_data *data = dev->data;
+
+	if (vbus_meas == NULL) {
+		return -EFAULT;
+	}
+
+	if (data->vbus_cb == NULL) {
+		return -EIO;
+	}
+
+	/* Call user supplied callback to measure VBUS */
+	*vbus_meas = data->vbus_cb(dev);
+
+	return 0;
+}
+
+/**
+ * @brief Enable or Disable Discharging of VBUS
+ *
+ * @return 0 on success
+ * @return -EIO on failure
+ */
+static int ucpd_set_discharge_vbus(const struct device *dev, bool enable)
+{
+	struct tcpc_data *data = dev->data;
+	int ret;
+
+	if (data->discharge_vbus_cb == NULL) {
+		return -EIO;
+	}
+
+	/* Call user supplied callback to discharge VBUS */
+	ret = data->discharge_vbus_cb(dev, enable);
+
+	return ret;
+}
+
+/**
+ * @brief UCPD interrupt init
+ */
+static void ucpd_isr_init(const struct device *dev)
+{
+	const struct tcpc_config *const config = dev->config;
+	struct tcpc_data *data = dev->data;
+	struct alert_info *info = &data->alert_info;
+
+	/* Disable all alert bits */
+	LL_UCPD_WriteReg(config->ucpd_port, IMR, 0);
+
+	/* Clear all alert handler */
+	ucpd_set_alert_handler_cb(dev, NULL, NULL);
+
+	/* Save device structure for use in the alert handlers */
+	info->dev = dev;
+
+	/* Initialize the work handler */
+	k_work_init(&info->work, ucpd_alert_handler);
+
+	/* Configure CC change alerts */
+	LL_UCPD_WriteReg(config->ucpd_port, IMR,
+			 UCPD_IMR_TYPECEVT1IE | UCPD_IMR_TYPECEVT2IE);
+	LL_UCPD_WriteReg(config->ucpd_port, ICR,
+			 UCPD_ICR_TYPECEVT1CF | UCPD_ICR_TYPECEVT2CF);
+
+	/* SOP'/SOP'' must be enabled via TCPCI call */
+	data->ucpd_rx_sop_prime_enabled = false;
+
+	stm32gx_ucpd_state_init(dev);
+
+	/* Configure and enable the IRQ */
+	config_tcpc_irq();
+}
+
+/**
+ * @brief Initializes the TCPC
+ *
+ * @return 0 on success
+ * @return -EIO on failure
+ */
+static int ucpd_init(const struct device *dev)
+{
+	const struct tcpc_config *const config = dev->config;
+	uint32_t cfg1;
+	int ret;
+
+	/* The UCPD port is disabled in the LL_UCPD_Init function */
+	ret = LL_UCPD_Init(config->ucpd_port,
```

Comment:
Just a comment: By using this function, you spare the need of defining in dts and setting clock using zephyr clock API.
This is not blocking, but having clocks from dts and using clock api would help in doing PM operation in future.

Reply:
I added a note that reflects your comment.